plan/stats.go

```diff
@@ -112,6 +112,9 @@ func (ds *DataSource) getStatsByFilter(conds expression.CNFExprs) *statsInfo {
 		}
 	}
 	ds.stats = profile
+	for _, colInfo := range ds.Columns {
```

Comment:
Wrap a method named something like `CollectColumns` may be better?

Reply:
But why do it here? Why not do it when estimating the row count?

Reply:
Because when estimating the row count, it's called by `HistColl` instead of `Table`.
https://github.com/pingcap/tidb/blob/master/statistics/table.go#L274 this check cannot be done by `HistColl`.

Reply:
The main reason to call `Invalid` is to load the needed column stats, so this for loop would make the load by need strategy useless.

Reply:
Columns there is after pruned, it does pruned some columns. If this pruning is not enough. Store the `tableid` in `expression.Column` will help?
Since we'll estimating the row count not only in the `datasource`, `TableID` cannot be maintained easily in `HistColl`. Store it in `expression.Column` may be more easier.

Reply:
What about only maintain the table id for the source table? We only needed to store it in table stats, rest of the code do not need to care about it, since, if the column is not needed in the data source, it won't be needed anymore.